tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ConfidentialClientIntegrationTests.cs

```diff
@@ -693,6 +698,51 @@ public async Task ClientCreds_ClientAssertion_AAD_NoWilson_Async()
             }
         }
 
+        [TestMethod]
+        public async Task ClientCreds_ServicePrincipal_OBO_PPE_Async()
+        {
+            X509Certificate2 cert = GetCertificate();
+            IReadOnlyList<string> scopes = new List<string>() { OBOServicePpeClientID + "/.default" };
+            IReadOnlyList<string> scopes2 = new List<string>() { OBOServiceDownStreamApiClientID + "/.default" };
+
+            var confidentialSPApp = ConfidentialClientApplicationBuilder
+                                    .Create(OBOClientPpeClientID)
+                                    .WithAuthority(PPEAuthenticationAuthority)
+                                    .WithCertificate(cert)
+                                    .WithTestLogging()
+                                    .Build();
+
+            var authenticationResult = await confidentialSPApp.AcquireTokenForClient(scopes).ExecuteAsync().ConfigureAwait(false);
+
+            string appToken = authenticationResult.AccessToken;
+            var userAssertion = new UserAssertion(appToken);
+            string atHash = userAssertion.AssertionHash;
+
+            var _confidentialApp = ConfidentialClientApplicationBuilder
+                .Create(OBOServicePpeClientID)
+                .WithCertificate(cert)
+                .Build();
+
+            var userCacheRecorder = _confidentialApp.UserTokenCache.RecordAccess();
+
+            authenticationResult = await _confidentialApp.AcquireTokenOnBehalfOf(scopes2, userAssertion)
+                                                         .WithAuthority(PPEAuthenticationAuthority)
+                                                         .ExecuteAsync().ConfigureAwait(false);
+
+            Assert.IsNotNull(authenticationResult);
+            Assert.IsNotNull(authenticationResult.AccessToken);
+
+            authenticationResult = await _confidentialApp.AcquireTokenOnBehalfOf(scopes2, userAssertion)
+                                                         .WithAuthority(PPEAuthenticationAuthority)
+                                                         .ExecuteAsync().ConfigureAwait(false);
+
+            Assert.IsNotNull(authenticationResult);
```

Comment:
Please add an Assert on authenticationResult.ResultMetadata.TokenSource - it should be from the cache.